conjecture-rust/src/engine.rs

```diff
@@ -95,8 +107,19 @@ impl MainGenerationLoop {
             Status::Overflow => (),
             Status::Invalid => self.invalid_examples += 1,
             Status::Valid => self.valid_examples += 1,
-            Status::Interesting => {
+            Status::Interesting(n) => {
                 self.best_example = Some(result.clone());
+                let mut changed = false;
+                self.minimized_examples.entry(n).or_insert_with(|| {result.clone()}); 
+                self.minimized_examples.entry(n).and_modify(|e| {
+                  if result < *e {
+                    changed = true;
+                    *e = result.clone()
+                  }; 
+                });
+                if changed {
+                  self.fully_minimized.remove(&n);
```

Comment:
Hmm, perhaps this is why the loop above is needed (instead of a condition)?

And to confirm my understanding: the loop must terminate, since each example can only ever shrink, and the size of the example is a non-negative integer.

Reply:
> Hmm, perhaps this is why the loop above is needed (instead of a condition)?

Right, exactly. The loop means that the following two things can occur:

1. We can discover other bugs during shrinking, so the number of distinct test cases can grow when shrinking a test case.
2. When we accidentally discover a better shrink for something else during our current shrink, that gives us a new starting point for the shrink process.

> And to confirm my understanding: the loop must terminate, since each example can only ever shrink, and the size of the example is a non-negative integer.

Almost but not quite. Examples shrink in what's called the shortlex order - either the size must decrease or it can stay the same and the value can go down lexicographically.

This still guarantees termination but not in a very useful way, because it might take exponentially many iterations to do so. In order to guard against this two things happen:

1. The shrinker is designed to avoid common ways of triggering exponential behaviour
2. There is a cap on the number of successful shrinks that can be performed before the engine stops running tests.

As it happens, neither of these are currently implemented in the Rust implementation of Conjecture! Fortunately in practice the issue doesn't come up too much, and they will be as part of the next milestone.

Reply:
Hmm one thing this points out is that I should really reflect the good commenting practice that the Python conjecture has in the Rust version. I'll add some commentary inline before merging.